pkg/storage/registry/dynamic/routingtree/routingtree.go

```diff
@@ -0,0 +1,174 @@
+// Copyright 2018-2023 CERN
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// In applying this license, CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+// Package routingtree contains the routing tree implementation.
+package routingtree
+
+import (
+	"errors"
+	"path"
+	"strings"
+)
+
+// Route represents a route inside a storage provider.
+type Route struct {
+	Path    string
+	MountID string
+}
+
+// RoutingTree is a tree made of maps of routes.
+type RoutingTree struct {
+	route Route
+	nodes map[string]*RoutingTree
+}
+
+// New returns a new RoutingTree.
+func New(rs []Route) *RoutingTree {
+	rt := &RoutingTree{
+		route: Route{
+			Path: "/",
+		},
+		nodes: make(map[string]*RoutingTree),
+	}
+
+	for _, r := range rs {
+		rt.AddRoute(r)
+	}
+
+	return rt
+}
+
+func (t *RoutingTree) addNode(r Route) *RoutingTree {
+	if t.route.Path == r.Path {
+		return t
+	}
+
+	newNode, ok := t.nodes[r.Path]
+	if !ok {
+		newNode = &RoutingTree{
+			route: r,
+			nodes: make(map[string]*RoutingTree),
+		}
+		t.nodes[r.Path] = newNode
+	}
+
+	return newNode
+}
+
+func getAncestors(p string) []string {
+	p = path.Clean(p)
+	ancestors := []string{}
+	previous := "/"
+	parts := strings.Split(p, "/")
+
+	for _, part := range parts {
+		previous = path.Join(previous, part)
+		ancestors = append(ancestors, previous)
+	}
+
+	return ancestors
+}
+
+// AddRoute adds a new Route to the RoutingTree based on a path `p`.
+func (t *RoutingTree) AddRoute(r Route) {
+	parts := getAncestors(r.Path)
+	current := t
+
+	for _, path := range parts {
+		newNode := Route{
+			Path: path,
```

Comment:
Supposing that the `RoutingTree` is storing the paths `/eos/user/g/gdelmont`, `/eos/project/c/cernbox`, `/cephfs/project/c/cernbox`, I would expect it to be in memory:
```
/
├── cephfs
│   └── project
│       └── c
│           └── cernbox
└── eos
    ├── project
    │   └── c
    │       └── cernbox
    └── user
        └── g
            └── gdelmont
```

instead of:

```
/
├── /cephfs
│   └── /cephfs/project
│       └── /cephfs/project/c
│           └── /cephfs/project/c/cernbox
└── /eos
    ├── /eos/project
    │   └── /eos/project/c
    │       └── /eos/project/c/cernbox
    └── /eos/user
        └── /eos/user/g
            └── /eos/user/g/gdelmont
```

The first I think is easier and you don't need to compute each time the parts from the `getAncestors` function for the lookup.